app/src/main/java/io/appium/settings/receivers/SmsReader.java

```diff
@@ -0,0 +1,110 @@
+
+public class SmsReader extends BroadcastReceiver implements HasAction {
+    private static final String TAG = SmsReader.class.getSimpleName();
+    private static final Uri INCOMING_SMS = Uri.parse("content://sms/inbox");
+    private static final String ACTION = "io.appium.settings.sms.read";
+    private static final int MAX_ITEMS = 100;
+    private static final String MAX_ITEMS_SETTING_NAME = "max";
+
+    private JSONObject listSms(Context context, int maxCount) throws JSONException {
+        Cursor cursor = context.getContentResolver().query(INCOMING_SMS,
+                null, null, null, "date desc");
+        //noinspection TryFinallyCanBeTryWithResources
+        try {
+            JSONArray items = new JSONArray();
+            if (cursor != null && cursor.moveToFirst()) {
+                do {
+                    JSONObject item = new JSONObject();
+                    for (String[] entry : new String[][]{
+                            {Telephony.Sms._ID, "id"},
+                            {Telephony.Sms.ADDRESS, "address"},
+                            {Telephony.Sms.PERSON, "person"},
+                            {Telephony.Sms.DATE, "date"},
+                            {Telephony.Sms.READ, "read"},
+                            {Telephony.Sms.STATUS, "status"},
+                            {Telephony.Sms.TYPE, "type"},
+                            {Telephony.Sms.SUBJECT, "subject"},
+                            {Telephony.Sms.BODY, "body"},
+                            {Telephony.Sms.SERVICE_CENTER, "serviceCenter"}
+                    }) {
+                        item.put(entry[1], formatJsonNull(cursor.getString(cursor.getColumnIndex(entry[0]))));
+                    }
+                    items.put(item);
+                } while (cursor.moveToNext() && items.length() < maxCount);
+            }
+            JSONObject result = new JSONObject();
+            result.put("items", items);
+            result.put("total", cursor == null ? 0 : cursor.getCount());
+            return result;
+        } finally {
+            if (cursor != null) {
+                cursor.close();
+            }
+        }
+    }
+
+    /**
+     * Responds to broadcast requests like
+     * am broadcast -a io.appium.settings.sms.read --es max 10
+     * with the list of the recent SMS messages formatted as JSON
+     */
+    @Override
+    public void onReceive(Context context, Intent intent) {
+        int maxItems = MAX_ITEMS;
+        if (intent.hasExtra(MAX_ITEMS_SETTING_NAME)) {
+            try {
+                maxItems = Integer.parseInt(intent.getStringExtra(MAX_ITEMS_SETTING_NAME));
+            } catch (NumberFormatException e) {
+                e.printStackTrace();
+            }
+        }
+        Log.d(TAG, String.format("Getting the recent %s SMS messages", maxItems));
+        String output;
+        try {
+            output = listSms(context, maxItems).toString();
+        } catch (Exception e) {
```

Comment:
How about show a message about the supported API version if the sdk version is lower than 19? 
Perhaps `Telephony.Sms.ADDRESS` etc need API level 19 https://developer.android.com/reference/android/provider/Telephony.TextBasedSmsColumns.html
(current minisdk is 18)

Reply:
it's strange why IDEA does not show these warnings. I'll add local constants instead

Reply:
👍